### PR TITLE
feat(user-attr-json): Add user attributes via tokens to portlet in la…

### PR DESCRIPTION
**Pull request by bjagg**
…youtDoc api

##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [x] tests are included
-   [x] documentation is changed or added
-   [x] view conforms with [WCAG 2.0 AA][]

##### Description of change

Need to support user values in URLs and other portlet values.

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties
Comment (ChristianMurphy) [inline gradle/wrapper/gradle-wrapper.properties]: what is the difference between the `-bin` and the `-all` distribution?

File: uPortal-layout/uPortal-layout-core/src/main/java/org/apereo/portal/layout/LayoutPortletPersonalizer.java
Comment (ChristianMurphy) [inline uPortal-layout/uPortal-layout-core/src/main/java/org/apereo/portal/layout/LayoutPortletPersonalizer.java]: Interesting, what are some use cases for customizing layout based on user attributes?

File: gradle/wrapper/gradle-wrapper.properties
Comment (bjagg) [reply gradle/wrapper/gradle-wrapper.properties]: Not sure -- I didn't update this file manually. I'll revert it.

File: uPortal-layout/uPortal-layout-core/src/main/java/org/apereo/portal/layout/LayoutPortletPersonalizer.java
Comment (bjagg) [reply uPortal-layout/uPortal-layout-core/src/main/java/org/apereo/portal/layout/LayoutPortletPersonalizer.java]: It's most often used in 2 cases. First, for setting an ID parameter in URLs for some non-sensitive services. Second, for personalization of text.

Comment (ChristianMurphy) [reply uPortal-layout/uPortal-layout-core/src/main/java/org/apereo/portal/layout/LayoutPortletPersonalizer.java]: makes sense, thanks for the context!

**Comment by ChristianMurphy** [comment]
when you have a chance could you run `./gradlew goJF` to format the code?

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java
Comment (jgribonvald) [inline uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java]: @bjagg you should review this part, @RequestParam on private method... an abusive copy/paste

**Comment by cbeach47** [comment]
> when you have a chance could you run `./gradlew goJF` to format the code?

Done!

**Comment by cbeach47** [comment]
> Why not using a bean conf injection to let peoples having the possibility to provide their own LayoutPortletPersonalizer ?
> 
> Else it's a great feature to customize layoutPortlet info !

@jgribonvald - good point - this is now in place.

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java]: fixed!

**Comment by cbeach47** [comment]
@jgribonvald / @bjagg / @ChristianMurphy - this is now ready for another review.
Details on the changes I implemented are in https://github.com/Jasig/uPortal/pull/1954/files#diff-b44ca7dde5c4bfff50889304f04ac8bae906d0095595072534876cbd9e3c198c

File: docs/implement/user-attributes/README.md
Comment (ChristianMurphy) [inline docs/implement/user-attributes/README.md:91]:
```suggestion
```properties
```

File: docs/implement/user-attributes/README.md
Comment (ChristianMurphy) [inline docs/implement/user-attributes/README.md:98]:
```suggestion
```properties
```

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java
Comment (ChristianMurphy) [inline uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java:35]: doc comment without any commentary?

File: uPortal-utils/uPortal-utils-core/src/main/java/org/apereo/portal/utils/personalize/PersonalizerImpl.java
Comment (ChristianMurphy) [inline uPortal-utils/uPortal-utils-core/src/main/java/org/apereo/portal/utils/personalize/PersonalizerImpl.java:31]: is this todo intended for this PR or a future one?

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/PortletsRESTController.java
Comment (jgribonvald) [inline uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/PortletsRESTController.java:106]: Why adding this line as it doesn't seem used ?

File: docs/implement/user-attributes/README.md
Comment (jgribonvald) [inline docs/implement/user-attributes/README.md:90]: Why enabling this feature by default ? I'm not sure that is required for everyone, it will run the feature even if no configuration is made, I'm not sure that's really good to run a code without the adopter configuration. 
Also if someone is interested he will enable it.

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java
Comment (jgribonvald) [inline uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java:118]: It may be better to refactor the code in another commit ?

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java
Comment (jgribonvald) [inline uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java:131]: It may be better to refactor the code in another commit ?

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java
Comment (jgribonvald) [inline uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java:114]: personalization with or without s ?

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java
Comment (jgribonvald) [inline uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java:119]: is it a good sentence ?

**Review by jgribonvald — COMMENTED** [review]
With @ChristianMurphy comments it seems good :+1:

File: docs/implement/user-attributes/README.md
Comment (bjagg) [reply docs/implement/user-attributes/README.md:90]: Yes, I would agree this should be disabled for backwards compatibility

Comment (cbeach47) [reply docs/implement/user-attributes/README.md:90]: sounds good, I'll flip it to disabled by default

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java:118]: commit or PR?  I could see a different commit for a cleaner review.  If the code refactor is associated with the code changes, I'd stick with the same PR.

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java:131]: plz see comment above

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/PortletsRESTController.java
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/PortletsRESTController.java:106]: good catch.  Leftover from a interim direction.  I'll remove it

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java:35]: whaaaaaa ?  :)  thx, provided some commentary

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java:114]: probably without the s.

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/utils/web/PersonalizationFilter.java:119]: I believe so.  Doesn't seem awkward and gets the point across...

File: uPortal-utils/uPortal-utils-core/src/main/java/org/apereo/portal/utils/personalize/PersonalizerImpl.java
Comment (cbeach47) [reply uPortal-utils/uPortal-utils-core/src/main/java/org/apereo/portal/utils/personalize/PersonalizerImpl.java:31]: interim direction.  I ended up using the http session object.  I'll remove the comment.

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java
Comment (jgribonvald) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/LayoutRESTController.java:118]: a commit is sufficient as in history, after merged, it won't be linked to the feature.

File: docs/implement/user-attributes/README.md
Comment (ChristianMurphy) [reply docs/implement/user-attributes/README.md:91]: @cbeach47 have you pushed the commit that include this change?
I'm not seeing it on GH yet :thinking:

**Review by ChristianMurphy — APPROVED** [review]

File: docs/implement/user-attributes/README.md
Comment (cbeach47) [reply docs/implement/user-attributes/README.md]: ah - realized this morning that I didn't pick up the README changes in that push.  I've pushed the README changes now.

**Pull request merged**
